compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeResolver.java

```diff
@@ -1450,7 +1450,7 @@ private void fillEffectiveType(BIntersectionType intersectionType,
         Iterator<BType> iterator = intersectionType.getConstituentTypes().iterator();
         BType effectiveType = iterator.next();
         BLangType bLangEffectiveType = bLangTypeItr.next();
-        if (effectiveType.tag == TypeTags.READONLY && iterator.hasNext()) {
+        if (Types.getImpliedType(effectiveType).tag == TypeTags.READONLY && iterator.hasNext()) {
```

Comment:
Here also better to exact this to variable and use that within the `calculateEffectiveType` function without doing that operation again in the `calculateEffectiveType` function.

Reply:
updated in [commit](https://github.com/ballerina-platform/ballerina-lang/pull/42025/commits/61e343e1616c463b2edaa6953ef0df4b986250d4).